### PR TITLE
Implement Mtom message encoding

### LINKED ISSUE #1810: Add Mtom support in WCF runtime
I am getting the following error when using the Visual Studio WCF Connected Service extension for Visual Studio 2017:

```
Scaffolding Code ...
Error:Warning: The following Policy Assertions were not Imported:   XPath://wsdl:definitions[@targetNamespace='http://imaging.oracle/']/wsdl:binding[@name='DocumentContentServicePortBinding']
  Assertions:
    <ns1:OptimizedMimeSerialization xmlns:ns1='http://schemas.xmlsoap.org/ws/2004/09/policy/optimizedmimeserialization'>..</ns1:OptimizedMimeSerialization>
Warning: Unsupported binding element type: 'System.ServiceModel.Channels.UnrecognizedAssertionsBindingElement'.
Warning: Unsupported message encoding element type: 'System.ServiceModel.Channels.MtomMessageEncodingBindingElement'. It must be one of the following types: 'System.ServiceModel.Channels.BinaryMessageEncodingBindingElement', 'System.ServiceModel.Channels.TextMessageEncodingBindingElement.'
Warning: Endpoint 'DocumentContentServicePort' at address 'http://localhost/imaging/ws/DocumentContentService' contains one or more bindings not compatible with .Net Core apps, skipping...
Error: No endpoints compatible with .Net Core apps were found.
Failed to generate service reference.
```

The WSDL/XSD file causing the issue:
[DocumentContentService.xsd.txt](https://github.com/dotnet/wcf/files/840688/DocumentContentService.xsd.txt)
[DocumentContentService.wsdl.txt](https://github.com/dotnet/wcf/files/840687/DocumentContentService.wsdl.txt)

dotnet --info
```
.NET Command Line Tools (1.0.0)

Product Information:
 Version:            1.0.0
 Commit SHA-1 hash:  e53429feb4

Runtime Environment:
 OS Name:     Windows
 OS Version:  6.1.7601
 OS Platform: Windows
 RID:         win7-x64
 Base Path:   C:\Program Files\dotnet\sdk\1.0.0
```

### PR DESCRIPTION
**Pull request by mconnew**
Fixes #1810

### CODE REVIEW

**Comment by mconnew** [comment]
Looks like tests failed because the test server had the wrong code installed. I suspect this might have been another test running in parallel. Trying again
/azp run

File: src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs [hunk 1]
```diff
@@ -0,0 +1,1553 @@
...
+                {
+                    _mimeWriter.StartPreface();
+                    _mimeWriter.WriteHeader(MimeGlobals.MimeVersionHeader, MimeGlobals.DefaultVersion);
+                    _mimeWriter.WriteHeader(MimeGlobals.ContentTypeHeader, _initialContentTypeForMimeMessage);
+                    _initialContentTypeForMimeMessage = null;
+                }
+
+                WriteMimeHeaders(_contentID, _initialContentTypeForRootPart, _isUTF8 ? MimeGlobals.Encoding8bit : MimeGlobals.EncodingBinary);
+
+                Stream infosetContentStream = _mimeWriter.GetContentStreamAsync().GetAwaiter().GetResult();
```
Comment (StephenMolloy) [inline src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs:142]: Why not just call the sync version of GetContentStream()? You haven't mixed sync/async in any other places.

File: src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs [hunk 2]
```diff
@@ -0,0 +1,1553 @@
...
+
+        internal void StartPreface()
+        {
+            if (state != MimeWriterState.Start)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.MimeWriterInvalidStateForStartPreface, state.ToString())));
+
+            state = MimeWriterState.StartPreface;
+        }
+
+        internal async Task StartPartAsync()
```
Comment (StephenMolloy) [inline src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs:1218]: Nit: Unlike all the other added async methods, this one along with FlushAsync and GetContentStreamAsync appear _before_ their synchronous counterparts.

**Review by StephenMolloy — APPROVED** [review]
A couple of questions. But I didn't see anything alarming. Looked quite a bit like NetFX code in most places. ;)

File: src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomReader.cs [hunk 3]
```diff
@@ -0,0 +1,3215 @@
...
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new XmlException(SR.Format(SR.MtomInvalidCIDUri, uri)));
+
+            if (_mimeParts != null && _mimeParts.TryGetValue(contentID, out part))
+            {
+                if (part.ReferencedFromInfoset)
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new XmlException(SR.Format(SR.MtomMimePartReferencedMoreThanOnce, contentID)));
+            }
+            else
+            {
+                // TODO: Fix this to be a configurable setting on MtomMessageEncodingBindingElement
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomReader.cs:427]: Future TODO as I also want to add some extensibility to enable controlling which elements get converted to a separate MimePart and which are inline. As they are both new api's I'll do them at the same time. I'll open an issue for this.

File: src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs (hunk 1 above)
Comment (mconnew) [reply src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs:142]: I originally made everything async, then hit this code so brought back the sync version to use here. Looks like I forgot to actually use it here. I'll fix this.

File: src/System.Private.ServiceModel/src/System/ServiceModel/Channels/TransportDefaults.cs [hunk 4]
```diff
@@ -17,7 +16,7 @@ namespace System.ServiceModel.Channels
     // If any of the const's are modified in this file, they must also be modified
     // on the Internal.ServiceModel.Primitives contract and implementation assemblies.
 
-    public static class EncoderDefaults
+    internal static class EncoderDefaults
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/System.Private.ServiceModel/src/System/ServiceModel/Channels/TransportDefaults.cs:19]: They shouldn't have been public in the first place. This is just some additional cleanup.

File: src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs (hunk 2 above)
Comment (mconnew) [reply src/System.Private.ServiceModel/src/Internals/System/Xml/XmlMtomWriter.cs]: Fixing, that is annoying 😄

File: src/System.Private.ServiceModel/src/Internals/System/Xml/XmlConverter.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Xml
+{
+    internal static class XmlConverter
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/System.Private.ServiceModel/src/Internals/System/Xml/XmlConverter.cs:7]: This is XmlConvert**er** and not XmlConvert. XmlConverter is internal in .NET, as is BinHexEncoder. I can't reference them from WCF.

**Review by HongGit — APPROVED** [review]

**Pull request merged**